### PR TITLE
operator/v1: add ingresscontroller

### PR DESCRIPTION
**Pull request by ironcladlou**
Introduce the `IngressController` API.

/cc @openshift/sig-network-edge @smarterclayton @deads2k 

TODO:
- [ ] Decide how to represent conditions (e.g. embed a common struct vs. bespoke)
- [ ] Make a decision about HostNetwork vs. NodePort

### CODE REVIEW

**Comment by ironcladlou** [comment]
Some outstanding questions:

1. Cluster vs. namespace scoped?
2. `v1` vs. `v1alpha1`?

**Comment by deads2k** [comment]
> 2\. `v1` vs. `v1alpha1`?

Our operators need to be v1. We support them and we aren't going to be able to break them.

**Comment by deads2k** [comment]
Need to update register.go.

**Comment by deads2k** [comment]
Is this the operator type used to drive the `Ingress` cluster operator?  Is there a non-ClusterIngress variant?  I expected this type to match the `co/ingress`.



> 1. Cluster vs. namespace scoped?

I see the following reasons to be namespace scoped.
1. it is a non-privileged action to create this
2. the thing it creates is logically scoped to a single namespace
3. multiple conflicting names in different namespaces is a thing this must support

I don't think any of those apply.  I would expect to see this cluster-scoped the same way the clusteroperator is cluster scoped.

**Comment by ironcladlou** [comment]
>Is this the operator type used to drive the `Ingress` cluster operator?

Maybe? Can you rephrase the question? It's the resource admins create/configure and that the operator acts upon to materialize routers and stuff.

>Is there a non-ClusterIngress variant? I expected this type to match the `co/ingress`.

Referring to [config/v1/types_ingress.go](https://github.com/openshift/api/blob/master/config/v1/types_ingress.go)? Can you explain a little more about your expectation?

**Comment by ironcladlou** [comment]
>I don't think any of those apply. I would expect to see this cluster-scoped the same way the clusteroperator is cluster scoped.

I also prefer the ergonomics of cluster scoped for this one (the namespacing seems to add only additional points of confusion), but we've gone back and forth on it before as a group. Let's see what others have to say.

**Comment by ironcladlou** [comment]
@smarterclayton PTAL

File: operator/v1/types_ingress.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterIngress describes a discrete, managed ingress component for the
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: might say this as

> ClusterIngress describes a managed ingress load balancer for the cluster. When a ClusterIngress is created a new load balancer is created to allow external traffic to reach the services that expose Ingress or Route resources.  Updating this resource may lead to disruption for public facing network connections as a new load balancer revision may be rolled out.

File: operator/v1/types_ingress.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterIngressSpec   `json:"spec,omitempty"`
+	Status ClusterIngressStatus `json:"status,omitempty"`
+}
+
+// ClusterIngressSpec is the specification of the desired behavior of the
+// ClusterIngress.
+type ClusterIngressSpec struct {
+	// IngressDomain is a DNS name suffix serviced by the ClusterIngress. If
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: All of your godoc for public API fields should use `ingressDomain` camelCase for field names.

File: operator/v1/types_ingress.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	// specified, this value:
+	//
+	// * Is added automatically to the status of Routes to inform downstream systems.
+	// * May be used for a wildcard DNS alias to any provisioned load balancer.
+	// * May be used to generate a default wildcard certificate.
+	//
+	// If nil, the default value from the cluster ingress config is used.
+	//
+	// Once set, this field cannot be changed.
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Why is this `*string` instead of `string`?

File: operator/v1/types_ingress.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	IngressDomain *string `json:"ingressDomain"`
+
+	// Replicas is the desired number of router instances.
+	Replicas int32 `json:"replicas"`
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Should be `*int32` and have the same rules as replica sets.  If unset, the ingress controller should manage that.

File: operator/v1/types_ingress.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+	// HighAvailability describes the kind of high availability mechanism to apply
+	// to the cluster ingress. If nil, a default appropriate for the
+	// infrastructure platform is used.
+	//
+	// Once set, this field cannot be changed.
+	HighAvailability *ClusterIngressHighAvailability `json:"highAvailability"`
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Can you describe for me what other values might be used?

File: operator/v1/types_ingress.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+	// DefaultCertificateSecret is the name of a secret containing the default
+	// certificate configuration. The secret is assumed to contain:
+	//
+	// * tls.crt
+	// * tls.key
+	DefaultCertificateSecret *string `json:"defaultCertificateSecret"`
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Should be `string` and empty means same as nil.

File: operator/v1/types_ingress.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+	// NamespaceSelector is used to filter the set of namespaces serviced by the
+	// router.
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector"`
+
+	// RouteSelector used to filter the set of routes serviced by the router.
+	RouteSelector *metav1.LabelSelector `json:"routeSelector"`
+
+	// NodePlacement enables explicit control over the scheduling of the router.
+	NodePlacement *NodePlacement `json:"nodePlacement"`
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Add description of default behavior.

File: operator/v1/types_ingress.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	// * tls.crt
+	// * tls.key
+	DefaultCertificateSecret *string `json:"defaultCertificateSecret"`
+
+	// NamespaceSelector is used to filter the set of namespaces serviced by the
+	// router.
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector"`
+
+	// RouteSelector used to filter the set of routes serviced by the router.
+	RouteSelector *metav1.LabelSelector `json:"routeSelector"`
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Add description of default behavior.

File: operator/v1/types_ingress.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	// DefaultCertificateSecret is the name of a secret containing the default
+	// certificate configuration. The secret is assumed to contain:
+	//
+	// * tls.crt
+	// * tls.key
+	DefaultCertificateSecret *string `json:"defaultCertificateSecret"`
+
+	// NamespaceSelector is used to filter the set of namespaces serviced by the
+	// router.
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector"`
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Add description of default behavior.

File: operator/v1/types_ingress.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+	// NoClusterIngressHA configures container-networked routers. Users are
+	// responsible for exposing the routers or connecting them to a load balancer.
+	NoClusterIngressHA ClusterIngressHAType = "None"
+)
+
+// ClusterIngressHighAvailability is the high availability configuration for a
+// ClusterIngress.
+type ClusterIngressHighAvailability struct {
+	Type ClusterIngressHAType `json:"type"`
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Needs godoc

Comment (smarterclayton) [reply operator/v1/types_ingress.go]: Godoc for this type should include the options (you need to describe what the user can set from within this comment)

File: operator/v1/types_ingress.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+// ClusterIngressStatus defines the observed status of the ClusterIngress.
+type ClusterIngressStatus struct {
+	// Replicas is the actual number of observed available router replicas
+	// according to the router deployment.
+	Replicas int32 `json:"replicas"`
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: AvailableReplicas

File: operator/v1/types_ingress.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+// ClusterIngressHighAvailability is the high availability configuration for a
+// ClusterIngress.
+type ClusterIngressHighAvailability struct {
+	Type ClusterIngressHAType `json:"type"`
+}
+
+// ClusterIngressStatus defines the observed status of the ClusterIngress.
+type ClusterIngressStatus struct {
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Add conditions and make sure you have an `Available` condition (at minimum).  I think that's the bar for GA

File: operator/v1/types_ingress.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
+package v1
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Be sure to set `+optional` on everything that is optional.

File: operator/v1/types_ingress.go (hunk 3 above)
Comment (smarterclayton) [reply operator/v1/types_ingress.go]: Don't say `nil`, this is godoc for JSON.  Say `if unset` or `if not specified`

Comment (smarterclayton) [reply operator/v1/types_ingress.go]: Describe the difference between null and empty.

Comment (smarterclayton) [reply operator/v1/types_ingress.go]: Also, explain the thing about if you don't set this for the next cluster ingress it won't work.

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: This field essentially enables a bundle of features:

1. publishing the value to route status ("router canonical hostname")
2. enables DNS management (when using a supported platform)
3. used as the basis for generating default wildcard certs as needed

We could have decoupled all those things (DNS seems inevitable and I started [a proposal for it](https://docs.google.com/document/d/1VuBowd1jTunXskScs8VXi8u-sND6THWuWcl1Gta2MOs/edit?usp=sharing) already)...

Is it valuable to be able to turn all of that off at once? That's the only possible meaning of an empty value today, and I'm not sure it makes any sense.

File: operator/v1/types_ingress.go (hunk 5 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: We've talked about this one again and don't think it needs to be a pointer. Empty string would mean use the default, and `None` would mean I don't want any

File: operator/v1/types_ingress.go (hunk 6 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: We've talked and do think this should be a value type and use empty string as an indicator to use defaults -- currently empty string means "use the cert baked into the image" which we don't think is useful

Comment (Miciah) [reply operator/v1/types_ingress.go]: @smarterclayton, would using `
DefaultCertificate *corev1.LocalObjectReference` for this field be acceptable?

File: operator/v1/types_ingress.go (hunk 3 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Spoke offline and decided this doesn't need to be a pointer after all. Will update docs.

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Looks better, updated

File: operator/v1/types_ingress.go (hunk 3 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Took another stab at this, added an outstanding question about how to cross-reference fields

File: operator/v1/types_ingress.go (hunk 4 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 12 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Should we be using `OperatorStatus` here or rolling our own?

File: operator/v1/types_ingress.go (hunk 13 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

File: operator/v1/types_ingress.go (hunk 11 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Changed

File: operator/v1/types_ingress.go (hunk 10 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

File: operator/v1/types_ingress.go (hunk 9 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

File: operator/v1/types_ingress.go (hunk 8 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

File: operator/v1/types_ingress.go (hunk 7 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

File: operator/v1/types_ingress.go (hunk 6 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: I changed to a reference

File: operator/v1/types_ingress.go (hunk 5 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

File: operator/v1/types_ingress.go (hunk 2 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

**Comment by ironcladlou** [comment]
@openshift/sig-network-edge @smarterclayton I think this is ready for another pass

File: operator/v1/types_ingress.go [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+	// * Used as the base domain of an automatically managed wildcard DNS record
+	//   on cloud platforms and when using Cloud highAvailability.
+	//
+	// * Used as the host for generated default wildcard certificates when no
+	//   explicit certificate is provided.
+	//
+	// ingressDomain must be unique among all ClusterIngress resources, and cannot
+	// be updated.
+	//
+	// TODO: What's the best way to reference another field in the API?
```
Comment (ironcladlou) [inline operator/v1/types_ingress.go]: Need some guidance here

**Comment by ironcladlou** [comment]
@openshift/sig-network-edge @smarterclayton major changes with the last commit.

* ClusterIngress -> IngressController
* HA types refactored
* `.spec.ingressDomain` -> `.spec.domain`

**Comment by ironcladlou** [comment]
This is ready for another round of review.

File: operator/v1/types_ingress.go [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+	// +optional
+	Domain string `json:"domain,omitempty"`
+
+	// replicas is the desired number of ingress controller replicas. If unset,
+	// defaults to 2.
+	//
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// highAvailability is used to expose the ingress controller and enable high
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: This is really platform balancer type or platform balancing strategy.

> What mechanism is used to take traffic from the DNS name to the ingress controller instances?

`Cloud` may be misnamed - is OpenStack a cloud?  I can use service load balancers and program DNS there.  Is RHHI a cloud?  They're going to have some variant of DNS and service load balancers.  

Why are we binding to the host instead of Node Ports in general? Why is NodePorts not an option?

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: I like PlatformBalancingStrategy.

We could expose the underlying Kube concept directly and it would be accurate: Cloud -> LoadBalancerService.

>Why are we binding to the host instead of Node Ports in general? Why is NodePorts not an option?

I assume because the use case involves bypassing the SDN hop into the cluster network. I don't see why NodePorts couldn't be an option but I suspect we still need Host. cc @knobunc

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Plus existing 3.x topologies are effectively `Host`, so even if perf isn't the concern it seems like we need the option for compatibility

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Is balancing even accurate in the general sense? PublishingStrategy?

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: From @Miciah on Slack with some other thoughts:
>PointOfPresence
>That's coming from the perspective that this option configures how far we're willing to run the line for you (to a public LB, to the service proxy, or no further than the pod host), but even if that's the right perspective, it's hard to come up with a common-use term that conveys it.
>TerminationPoint, ManagementEndpoint...

Comment (smarterclayton) [reply operator/v1/types_ingress.go]: PointOfPresence is really obscure.

PlatformBalancingStrategy is good.  EndpointPublishingStrategy is good.  Slightly prefer publishing to balancing (this is where you publish the endpoints for the ingress controller).

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Ran with `EndpointPublishingStrategy`.  PTAL

File: operator/v1/types_ingress.go [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+	Domain string `json:"domain,omitempty"`
+
+	// replicas is the desired number of ingress controller replicas. If unset,
+	// defaults to 2.
+	//
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// endpointPublishingStrategy is used to publish the ingress controller
+	// endpoints to other netowrks, enable load balancer integrations, etc.
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: typo

File: operator/v1/types_ingress.go [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+	//
+	// If empty, the default is based on the cluster platform:
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Enumerate all possible strategies in this comment (the godoc on the constant doesn't show up in oc explain)

File: operator/v1/types_ingress.go [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+	//
+	// +optional
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+}
+
+// EndpointPublishingStrategyType is a way to publish ingress controller endpoints.
+type EndpointPublishingStrategyType string
+
+const (
+	// LoadBalancerService publishes the ingress controler using a Kubernetes
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: typo

File: operator/v1/types_ingress.go [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+
+	// Private does not publish the ingress controller.
+	//
+	// In this configuration, the ingress controller deployment uses container
+	// networking, and is not explicitly published. The user must manually publish
+	// the ingress controller.
+	PrivateStrategyType = "Private"
+)
+
+// EndpointPublishingStrategy is the a way to publish the endpoints of an
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: typo

File: operator/v1/types_ingress.go [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+// IngressController, and represents the type and any additional configuration
+// for a specific type.
+type EndpointPublishingStrategy struct {
+	// type is the publishing strategy to use. Valid values are
```
Comment (smarterclayton) [inline operator/v1/types_ingress.go]: Explain them in this comment (instead of the other comment)

File: operator/v1/types_ingress.go (hunk 17 above)
Comment (smarterclayton) [reply operator/v1/types_ingress.go]: Actually, don't, do it on `type`.

**Comment by smarterclayton** [comment]
Should be `v1`.  All our public APIs will be locked forever once we ship.

File: operator/v1/types_ingress.go (hunk 17 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

File: operator/v1/types_ingress.go (hunk 12 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: @smarterclayton still need to know what to do here (define our own conditions vs. embedding some common struct)

File: operator/v1/types_ingress.go (hunk 16 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 18 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 19 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 20 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

**Comment by ironcladlou** [comment]
* Shortened the IngressController doc by moving info closer to the relevant fields.

* Reframe the DNS management behavior in terms of the LoadBalancerService publishing strategy which I hope makes more sense.

* Tried to clarify the certificate behavior.

* Tried to further clarify domain.

PTAL, I'm much happier with this rev, and am getting a lot of use from the feedback

File: operator/v1/types_ingress.go [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+	//
+	// The secret must contain the following keys and data:
+	//
+	//   tls.crt: certificate file contents
+	//   tls.key: key file contents
+	//
+	// If unset, a wildcard certificate is automatically generated and used. The
+	// certificate is valid for the ingress controller domain (and subdomains) and
+	// the generated certificate's CA will be automatically integrated with the
+	// cluster's trust store.
```
Comment (ironcladlou) [inline operator/v1/types_ingress.go]: @enj what's the best way to refer to "the cluster's trust store" here?

File: operator/v1/types_ingress.go [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+	//
+	// The in-use certificate (whether generated or user-specified) will be
+	// automatically integrated with OpenShift's built-in OAuth server.
```
Comment (ironcladlou) [inline operator/v1/types_ingress.go]: @enj what's the best way to refer to "OpenShift's built-in OAuth server" here?

File: operator/v1/types_ingress.go [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+	// Publishes the ingress controller using a Kubernetes LoadBalancer Service.
+	//
+	// In this configuration, the ingress controller deployment uses container
+	// networking. A LoadBalancer Service is created to publish the deployment.
+	//
+	// See: https://kubernetes.io/docs/concepts/services-networking/#loadbalancer
+	//
+	// If domain is set, a wildcard DNS record will be managed to point at the
+	// LoadBalancer Service's external name. DNS records are managed only in DNS
+	// zones defined in the DNS cluster configuration resource.
```
Comment (ironcladlou) [inline operator/v1/types_ingress.go]: @deads2k what's the best way to refer to the `dns.config.openshift.io/cluster` resource here?

File: operator/v1/types_ingress.go [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+	// * When using a generated default certificate, the certificate will be valid
+	//   for domain and its subdomains. See defaultCertificate.
+	//
+	// * The value is published to individual Route statuses so that end-users
+	//   know where to target external DNS records.
+	//
+	// domain must be unique among all IngressControllers, and cannot be
+	// updated.
+	//
+	// If empty, defaults to the cluster Ingress config domain value.
```
Comment (ironcladlou) [inline operator/v1/types_ingress.go]: @deads2k what's the best way to refer to the `ingress.config.openshift.io/cluster` resource here?

File: operator/v1/types_ingress.go (hunk 23 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 24 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: fixed

**Comment by smarterclayton** [comment]
We can iterate in follow up.

/lgtm
